### ISSUE TITLE
Devtools flamegraph component sizes are not proportional to render time

### ISSUE THREAD

**Issue by pimterry** [facebook/react#16596]
**Do you want to request a *feature* or report a *bug*?**: Bug

**What is the current behavior?**
The flamegraph appears to be intended to show the component tree with each component sized proportional to the time taken to render it & its children. Either it doesn't do this reliably, or the relative sizes are intended to show something else (in which case: I don't know what). As an example:

![Screenshot from 2019-08-28 13-53-00](https://user-images.githubusercontent.com/1526883/63853615-7580cd00-c99b-11e9-83b9-0a53d1d954c6.png)

Compare the sizes of `pane1` & `pane2` (the two direct children of the highlighted SplitPane). Pane1 is approx 75% of the size, and took 3.7ms to render, whilst Pane2 is 25% of the size and took 23.2ms.

This UI encourages me to focus on Pane1, but Pane2 is definitely what I should be looking at.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem:**

I'm not sure of the steps to reproduce this in general, but it's completely reliable when profiling my application in quite a few places, I think this is likely a general issue. Not sure, but I suspect comparing the numbers and relative sizes when profiling any substantial render will find odd mismatches.

Let me know if that doesn't seem to be true, and I'll find time to do some more testing & build a proper repro.

**What is the expected behavior?**

The components should be sized relative to their overall render time, so you can easily find the most expensive component trees.

It's totally possible that the bug here is that my expectations are wrong, in which case I'd love to know what this is _supposed_ to show, and imo it'd be worth considering showing the proportional render time instead.

**Which versions of React, and which browser / OS are affected by this issue?**

React: 16.6.3
Chrome: 75.0.3770.142 
React devtools addon: `Created from revision a39d9c3 on 8/26/2019.`

**Comment by pimterry** [comment]
Ah, interesting! Thanks @bvaughn, that's very helpful. I understand it might not be everybody's preference, but personally I would find the option 1 you suggested much more useful for the kind of profiling I'm doing. The overall time taken to render a tree if I do it from scratch isn't important to me here: I want to know why _this_ render was slow.

Perhaps there's an argument for 'show all'/'show only rendered' toggle on the flamegraph? Although adding more toggles does feel like a bit of a cop out.

> > The components should be sized relative to their overall render time, so you can easily find the most expensive component trees.
> 
> Color is used to convey this information, as mentioned in the blog post above.

> [from the blog post:] The color of a bar indicates how long the component (and its children) took to render in the selected commit.

Is this definitely correct? In my screenshot above, it seems to be clearly coloured by how long the single component took to render, totally ignoring its children. If this is correct, I don't understand why pane2 isn't glowing orange.

Colouring ignoring children like this means that the colour doesn't help sadly. It highlights individually expensive single components, but doesn't help for trees that are cumulatively expensive.

More generally: in the screenshot above, it would be useful if it was obvious somehow that pane2 was cumulatively very expensive to render. In fact with a small tweak I can avoid rendering that entirely in many cases, which will shorten this render by 80%+. It's not clear from the graph shown that that's where I should be investigating at all, and I'd find it very useful if it was.

**Comment by pimterry** [comment]
> You may find the "Ranked" view a little more to your liking? Although it doesn't preserve the tree structure so maybe it's not very helpful.

I think the 'Ranked' view also shows the individual cost too, not cumulative cost, is that not right? Also yes, the tree structure is lovely and very helpful, and it'd be nice to have that :smiley:.

I do find the ranking useful for chasing down individually expensive components occasionally, which I definitely do want at times, but personally most of my profiling tends to be focused on avoiding rendering things at all, not worrying about the cost of an individual render() call.

> The combination of color and width help here. (Also the labels!) In the portion of your screenshot that you shared, the profiler shows that a big chunk of your app re-rendered, but no single component was very expensive. This is useful information. Zoomin in on e.g. SplitPane we can see that its subtree took 27.2ms.

Yep, the labels definitely have the info I'm looking for. There's no clear signal on which labels to look at though; you'd have to just read all of them to work out where the key place to trim the tree is, as it were, which isn't great. In this case there's the one main branch there, but it would be easy to get harder to spot cases.

Meanwhile, take a look at my artfully re-proportioned version of that the same profile:

![fixed-Screenshot from 2019-08-28 13-53-00](https://user-images.githubusercontent.com/1526883/63872752-9575b800-c9be-11e9-986a-856182ad2d19.png)

That matches the actual time spent in this render. Here it's pretty clear that there's big savings available if you can skip updating pane2, and pane1 probably isn't worth focusing on right now.

> > Perhaps there's an argument for 'show all'/'show only rendered' toggle on the flamegraph?
> 
> You can always make an argument, but I don't think you could convince me to build it :grin:

How about if _I_ build it - could I persuade you to merge it? :grin:

**Comment by pimterry** [comment]
> To be clear, all of the coloring is self-duration only, not child duration. That's because coloring with child duration is essentially pointless.

Yep, I can totally see how cumulative colouring wouldn't be useful, makes sense!

> I don't think it would be a useful UX.

Ok, but there's a specific question I'm trying to ask while profiling, which I think is a useful & common question, that the current profiler doesn't easily answer: how _cumulatively_ expensive was each part of _this specific commit_. Right now:

Node width = cumulative expense within a _full-tree_ commit
Colour = _individual_ expense within this specific commit

The only place to see cumulative within this specific commit is on the labels, which are not easy to skim through.

Adding a 'only rendered this commit' toggle lets the node width switch from showing the full tree to showing just this commit, which gives the info I'm looking for.

When you say you don't think this would be useful, is it that you don't think this is a sensible question to ask, or you don't think this will answer it, or it is not useful for some other reason?

> How would a flame graph look if two (or more) disconnected subtrees rendered?

I think the root of the tree could still be shown as now, even if it hasn't rendered, as that doesn't need to affect the width of the nodes that rendered beneath it.

The issue is the non-rendered child nodes, which are defining the width of their rendered parents, even though they're not involved in this commit. I'd just want to hide those. Keeping the tree root would also keep the basic shape of the tree more stable - it just gets shorter/taller, with much of the same overall structure otherwise.

It might be possible to solve this not by hiding them but by showing their width differently somehow? I can't see an easy route there though, lots of edge cases that end up with non-rendered children all pushed into a tiny sliver that seems broken.

**Comment by bvaughn** [comment]
This is an information dense graph. It's hard to represent all possible information at a top level glance. Some of it requires digging in a little. I think the current flame graph makes a reasonable set of trade offs to maximize readability when comparing commits (e.g. "what changed from the previous render to this one") while still making it possible to see where the expensive component(s) lie (color). I agree that spotting an expensive subtree requires a bit of a mental join, but I haven't so far heard a proposal that I think wouldn't result in a confusing UX when stepping between commits.

> I think the root of the tree could still be shown as now, even if it hasn't rendered, as that doesn't need to affect the width of the nodes that rendered beneath it.

I don't think this would work in all cases. What if one of the parts is a deep descendant of the other. What do we render in between?

> It might be possible to solve this not by hiding them but by showing their width differently somehow?

I don't think this would be possible without adding visual confusion to the graph. One nice thing about the flame graphs now is that stepping through commits provides a nice visual hint about what changed. The eye is drawn to size and/or color changes. Those things changed. If other things changed too, I think it might add a lot of noise.

This discussion is getting pretty long and I'm probably going to have to back out for now and focus on some other things. Hope that's okay!
